### PR TITLE
FEAT: Cleanup directory based on duplicate file

### PR DESCRIPTION
**Pull request by yahayaohinoyi**
https://github.com/users/yahayaohinoyi/projects/4?pane=issue&itemId=79817740

## CONTEXT
We want to clean directories based on files that have similar content

## DESCRIPTION
Presently, I'm using `first_string_in_file_name_before_whitespace + filesize` as the key to tell if a file is similar to another. This is the simplest way I could think of.

## HOW THIS WAS TESTED
`Unit tests`
`Manually`

- Create a folder with 3 different files but the same prefix before the space character; `data/test 1.txt`, `data/test 2.txt`, `data/test 2.txt`
- Ensure the content of two of these files are same
- run the code and notice one of the files with same content got deleted

### CODE REVIEW

File: src/features/cleaner_file_duplicate.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+                    // use first word in filename + filesize as they key till we think of something
+                    // better
+
+                    let file_key = get_file_key(path.file_name(), metadata.len());
+
+                    match file_key {
+                        Some(file_key) => {
+                            if set.contains(&file_key) {
+                                if !dry_run {
```
Comment (mubarak-ramadan) [inline src/features/cleaner_file_duplicate.rs]: This block is duplicated everywhere in the codebase, can we extract this out into a separate method and call it anywhere instead ?

File: src/features/cleaner_file_duplicate.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+pub fn directory_cleaner_based_on_duplicate_files(directory: &String, dry_run: bool) -> Result<()> {
+    let mut set: HashSet<String> = HashSet::new();
+    for entry in WalkDir::new(directory).into_iter() {
+        match entry {
+            Ok(dir) => {
+                let path = dir.path();
+
+                if path.is_file() {
+                    let metadata = fs::metadata(path)
```
Comment (mubarak-ramadan) [inline src/features/cleaner_file_duplicate.rs:24]: same here, maybe create an util class, and we can just do getMetadata. etc

**Review by mubarak-ramadan — COMMENTED** [review]
just minor comments

File: src/features/cleaner_file_duplicate.rs (hunk 1 above)
Comment (yahayaohinoyi) [reply src/features/cleaner_file_duplicate.rs]: I could do this as it feels similar enough. The code block doesn't seem much in my opinion but not against it

File: src/features/cleaner_file_duplicate.rs (hunk 2 above)
Comment (yahayaohinoyi) [reply src/features/cleaner_file_duplicate.rs:24]: so getMetadata would be `fs::metadata(path).with_context(|| format!("Failed to read metadata for file: {:?}", path))?;`. Doesn't make it cleaner IMO. That's a small enough code blob

**Review by mubarak-ramadan — APPROVED** [review]

**Pull request merged**